keps/sig-architecture/2019-07-16-node-role-label-use.md

```diff
@@ -0,0 +1,215 @@
+
+## Motivation
+
+The `node-role.kubernetes.io/master` (and the broader `node-role.kubernetes.io` namespace for labels) was introduced to provide a simple organizational and grouping convention for cluster users. The labels were reserved solely for organizing nodes via a convention that tools could recognize to display information to end users, and for use by opinionated external tooling that wished to simplify conventions for end users. Use of the label by components within the Kubernetes project (those projects subject to API review) was expressely forbidden. Specifically, no project could mandate the use of those labels in a conformant distribution, since we anticipated that many deployments of Kubernetes would have more nuanced control-plane topologies than simply "a control plane node".
+
+Over time, several changes to Kubernetes core and related projects were introduced that depended on the `node-role.kubernetes.io/master` label to vary their behavior in contravention to the guidance the label was approved under. This was unintentional and due to unclear reviewer guidelines that have since been more strictly enforced. Likewise, the complexity of Kubernetes deployments has increased and the simplistic mapping of control plane concepts to a node has proven to limit the ability of conformant Kubernetes distributions to self-host, as anticipated.
+
+
+### Goals
+
+This KEP:
+
+* Clarifies that the use of the `node-role.kubernetes/*` label namespace is reserved solely for end-user and external Kubernetes consumers, and:
+  * May not be used to vary behavior within Kubernetes projects that are subject to API review (kubernetes/kubernetes and all components that expose APIs under the `*.k8s.io` namespace)
+  * Is not required to be used for a cluster to be conformant
+* Describes the locations within Kubernetes that must be changed to use an alternative mechanism for behavior
+  * Suggests approaches for each
+* Describes the timeframe and migration process for Kubernetes distributions and deployments to update labels
+
+
+## Proposal
+
+### Use of `node-role.kubernetes.io/*` labels
+
+* Kubernetes components MUST NOT set or alter behavior on any label within the `node-role.kubernetes.io/*` namespace.
+* Kubernetes components (such as `kubectl`) MAY simplify the display of `node-role.kubernetes.io/*` labels to convey the node roles of a node
+* Kubernetes examples and documentation MUST NOT leverage the node-role labels for node placement
+* External users, administrators, conformant Kubernetes distributions, and extensions MAY use `node-role.kubernetes.io/*` without reservation
+  * Extensions are recommended not to vary behavior based on node-role, but MAY do so as they wish
+* First party components like `kubeadm` MAY use node-roles to simplify their own deployment mechanisms.
```

Comment:
@smarterclayton as of today kubeadm use `node-role.kubernetes.io/master` for labeling control-plane nodes so the users can later have control about workload placement, while such label is not driving/impacting kubeadm deployment mechanisms.
Giving this fact, should we consider kubeadm as already compliant to this KEP or do you expect kubeadm to change somehow?

cc @neolit123 ^^

Reply:
i think, the "Motivation" section gives the answer.
kubeadm and similar deployers outside of core should be fine to use `node-role.kubernetes.io`.
`node-role.kubernetes.io` will continue to be a namespace that triggers special behavior in kubectl.

it's unfortunate that's it's called `/master` in kubeadm still, yet changing it can be quite breaking and needs a GA+ deprecation period, action required and having another `node-role` in parallel during the deprecation (multiple `node-roles` are technically supported).

Reply:
`kubeadm` is allowed to use role labels because it is an opinionated deployment tool.  But for instance because `kubeadm` is allowed to use the label doesn't mean that `kops` or `gke` are required to use it or set it in order to get kube function, as you note.

I didn't want to get into naming, that should be a separate issue.